zookeeper/src/main/java/com/linecorp/armeria/client/zookeeper/ZooKeeperEndpointGroup.java

```diff
@@ -67,58 +70,80 @@ public ZooKeeperEndpointGroup(String zkConnectionStr, String zNodePath, int sess
-     * @return  {@link ZooKeeperListener}
+     * Create a ZooKeeper-based {@link EndpointGroup}, endpoints will be retrieved from a node's all children's
+     * node value using {@link NodeValueCodec}.
+     * @param client the curator framework instance.
+     * @param zNodePath       a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param nodeValueCodec  the nodeValueCodec
      */
-    private ZooKeeperListener createListener() {
-        return new ZooKeeperListener() {
-            @Override
-            public void nodeChildChange(Map<String, String> newChildrenValue) {
-                final List<Endpoint> newData = newChildrenValue.values().stream()
-                                                               .map(nodeValueCodec::decode)
-                                                               .filter(Objects::nonNull)
-                                                               .collect(toImmutableList());
-                final List<Endpoint> prevData = endpoints();
-                if (!prevData.equals(newData)) {
-                    setEndpoints(newData);
-                }
-            }
-
-            @Override
-            public void nodeValueChange(String newValue) {
-                //ignore value change event
-            }
-
-            @Override
-            public void connected() {
+    public ZooKeeperEndpointGroup(CuratorFramework client, String zNodePath, NodeValueCodec nodeValueCodec) {
+        checkArgument(!isNullOrEmpty(zNodePath), "zNodePath");
+        this.nodeValueCodec = requireNonNull(nodeValueCodec, "nodeValueCodec");
+        this.internalClient = false;
+        this.client = requireNonNull(client);
+        client.start();
+        pathChildrenCache = new PathChildrenCache(client, zNodePath, true);
+        pathChildrenCache.getListenable().addListener((c, event) -> {
+            switch (event.getType()) {
+                case CHILD_ADDED:
+                    addEndpoint(this.nodeValueCodec.decode(event.getData().getData()));
+                    break;
+                case CHILD_REMOVED:
+                    removeEndpoint(this.nodeValueCodec.decode(event.getData().getData()));
+                    break;
+                default:
+                    break;
             }
-        };
-    }
-
-    @VisibleForTesting
-    void enableStateRecording() {
-        zooKeeperConnector.enableStateRecording();
-    }
-
-    @VisibleForTesting
-    ZooKeeper underlyingClient() {
-        return zooKeeperConnector.underlyingClient();
+        });
+        try {
+            pathChildrenCache.start();
+        } catch (Exception e) {
+            throw new IllegalStateException(e);
+        }
     }
 
-    @VisibleForTesting
-    BlockingQueue<KeeperState> stateQueue() {
-        return zooKeeperConnector.stateQueue();
+    @Override
+    public void close() {
+        try {
+            pathChildrenCache.close();
+            if (internalClient) {
+                client.close();
```

Comment:
- Please run this in the finally block so that `client.close()` is called even if `pathChildrenCache.close()` raises an exception.
- Do `try-catch` on `client.close()` and do `logger.warn("Failed to close CuratorFramework:", e)`